File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-10.cpp
class Solution {
public:
    unordered_map<int, unordered_map<int, int>> dp;
    int maxProfit(vector<int>& prices) {
        return dfs(0, 1, prices);
    }
    int dfs(int i, int canbuy, vector<int>& prices) {
        if (i >= prices.size()) {return 0;}
        if (dp.count(i) and dp[i].count(canbuy)) {
            return dp[i][canbuy];
        }
        int cooldown = dfs(i + 1, canbuy, prices);
        int profit;
        // 1
        if (canbuy) {
            profit = dfs(i + 1, 0, prices) - prices[i];
        } else {
            profit = dfs(i + 2, 1, prices) + prices[i];
        }
        dp[i][canbuy] = max(cooldown, profit);
        return dp[i][canbuy];
    }
};

File: Data Structures & Algorithms/longest-consecutive-sequence/submission-10.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        nums = set(nums)
        ans = 0
        for n in nums:
            if n - 1 in nums:
                continue
            curr = n
            count = 1
            while curr + 1 in nums:
                count += 1
                curr += 1
            ans = max(count, ans)
        return ans